File: python/taflow/indicators/volume_zone_oscillator.py
"""Public adapter for the native Volume Zone Oscillator state."""

from typing import Any

import numpy as np

from .._native import VolumeZoneOscillator as _Native
from .._series import as_float64_series


class VolumeZoneOscillator:
    """Measure EMA-smoothed signed volume as a percentage of total volume.

    Volume is signed from each close change, then the signed and unsigned
    streams are independently EMA-smoothed. The first close establishes
    direction, so output remains ``NaN`` for ``timeperiod`` additional bars.
    This definition maps to Wickra ``VZO`` 0.9.9; TA-Lib has no equivalent.

    Args:
        close: Initial chronological closing-price series.
        volume: Initial chronological volume series aligned with ``close``.
        timeperiod: EMA smoothing period. Defaults to 14.

    Raises:
        ValueError: If the input lengths differ or ``timeperiod`` is zero.
    """

    def __init__(self, close: Any, volume: Any, timeperiod: int = 14) -> None:
        """Initialize the oscillator and process aligned close/volume history."""
        self._state = _Native(int(timeperiod))
        self.extend(close, volume)

    def append(self, close: float, volume: float) -> "VolumeZoneOscillator":
        """Append one close/volume bar and return this instance."""
        self._state.append(float(close), float(volume))
        return self

    def extend(self, close: Any, volume: Any) -> "VolumeZoneOscillator":
        """Append aligned close and volume series.

        Returns:
            This instance, allowing method chaining.

        Raises:
            ValueError: If ``close`` and ``volume`` have different lengths.
        """
        close_values = as_float64_series(close)
        volume_values = as_float64_series(volume)
        if len(close_values) != len(volume_values):
            raise ValueError("close and volume must have equal lengths")
        self._state.extend(close_values, volume_values)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest oscillator value, or ``None`` during warm-up."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned oscillator values, including warm-up ``NaN``."""
        return self._state.compute()

    def reset(self) -> "VolumeZoneOscillator":
        """Clear close direction and EMA state, then return this instance."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of aligned bars processed by Rust."""
        return len(self._state)


__all__ = ["VolumeZoneOscillator"]

File: crates/taflow-core/src/indicators/volume_zone_oscillator.rs
//! Incremental Volume Zone Oscillator.

use crate::error::TaResult;
use crate::indicators::ExponentialMovingAverage;
use crate::stream::StreamingIndicator;

/// EMA-smoothed signed volume as a percentage of EMA-smoothed volume.
///
/// The first close establishes direction. Later volume is positive, negative,
/// or zero when close rises, falls, or remains unchanged. Both EMA states use
/// an SMA seed, matching Wickra `VZO` 0.9.9 and producing the first value after
/// `timeperiod + 1` bars.
#[derive(Debug, Clone)]
pub struct VolumeZoneOscillator {
    signed_volume_average: ExponentialMovingAverage,
    volume_average: ExponentialMovingAverage,
    previous_close: Option<f64>,
    value: Option<f64>,
}

impl VolumeZoneOscillator {
    /// Creates the oscillator with a positive EMA smoothing period.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        Ok(Self {
            signed_volume_average: ExponentialMovingAverage::new(timeperiod)?,
            volume_average: ExponentialMovingAverage::new(timeperiod)?,
            previous_close: None,
            value: None,
        })
    }

    /// Appends one close/volume bar and returns the latest warm value.
    pub fn append(&mut self, close: f64, volume: f64) -> Option<f64> {
        let Some(previous_close) = self.previous_close.replace(close) else {
            self.value = None;
            return None;
        };

        let signed_volume = if close > previous_close {
            volume
        } else if close < previous_close {
            -volume
        } else {
            0.0
        };
        let signed_average = self.signed_volume_average.append(signed_volume);
        let volume_average = self.volume_average.append(volume);

        self.value = signed_average.zip(volume_average).map(|(signed, total)| {
            if total == 0.0 {
                0.0
            } else {
                100.0 * signed / total
            }
        });
        self.value
    }

    /// Returns the latest oscillator value, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Clears direction, both EMA states, and the latest value.
    pub fn reset(&mut self) {
        self.signed_volume_average.reset();
        self.volume_average.reset();
        self.previous_close = None;
        self.value = None;
    }
}
